Write a blog post about the test framework (#56)
invictus-integration/testing-framework: blog/invictus-logic-apps-testing-framework.md

Announcing the Invictus Test Framework for Azure Logic Apps
--

> >While writing unit or integration tests should be part of every project, this proofs to be quite difficult whenever creating integration scenarios using Azure Logic Apps.

Reviewer: Very nice introduction! Even I as a LogicApp-dummy understand it 😅 !

>Thanks to the creation of the Invictus Test Framework for Azure Logic Apps, there is no longer any excuse not to write any test-cases.

> Up until now, when building interfaces on top of Azure Logic Apps, you were forced to manually (re)test every implemented feature because there simply wasn't anything available to help you on your way, which ends up costing both time and money you'd like to spend elsewhere.  

Reviewer 2: ```suggestion
Up until now, when building interfaces on top of Azure Logic Apps, you were forced to manually (re)test every implemented feature because there simply wasn't anything available to help you on your way.

While this approach works fine, we've found that it has a few flaws:
- New people on your team don't have an easy way to understand how flows should work
- It's cumbersome since you cannot easily automate it to guarantee quality
- It takes time & effort which ends up in a higher engineering & testing cost, money you'd like to spend elsewhere
```

> Since developing the Invictus Methodology - which includes patterns, best practices and several templates - Codit has been investing in building a solid foundation enabling us to quickly kick off new projects, while ensuring a high level of quality. The biggest thing which, until now, was still lacking, was the ability to include test-cases into those projects.  

Reviewer 2: ```suggestion
Since developing the Invictus Methodology, which includes patterns, best practices and several templates, Codit has been investing in building a solid foundation enabling us to quickly kick-off new projects, while ensuring a high level of quality. The biggest thing which, until now, was still lacking, was the ability to include test-cases into those projects.  
```

> Because of this, we started working on a set of classes allowing us to easily enable/disable, trigger and even modify existing Azure Logic Apps. But, since the operations are not enough to write full-blown test-scenarios, we extended this list to ensure you would also be able to monitor the executed Logic App-runs.

Reviewer 2: ```suggestion
Because of this, we started working on a framework that allowed us to easily enable/disable, trigger and even modify existing Azure Logic Apps. But, since the operations are not enough to write full-blown test-scenarios, we extended this list to ensure you would also be able to monitor the executed Azure Logic App runs.
```

> All these operations have now been forged into a test framework for Azure Logic Apps, which has been made available on NuGet and can be included in your project using the following command:

Reviewer 2: ```suggestion
All these operations have now been forged into the Invictus Test Framework for Azure Logic Apps which we used internally for our customers and have proven to be effective.

Today, **we are happy to announce that we are open-sourcing Invictus Test Framework for Azure Logic Apps on GitHub** and it is now available on NuGet for you to use on your projects!
```

Author: Wasn't sure whether this was really open-sourced or just but onto GitHub and made available (code-wise) to Invictus customers :)

> ```shell

Reviewer 2: ```suggestion
Get started very easily :
```shell
```

But what exactly do you need to specify in order to really authenticate?  
Well, below is an overview of the required information:

```csharp
// The Tenant ID as found in the AAD properties
string tenantId = "my-tenant-id";
// The ID of the subscription containing your Logic Apps.
string subscriptionId = "my-subscription-id";
// The Client ID of the service principal to be used to authenticate.
string clientId = "my-client-id";
// The Client Secret created along with the service principal used for authentication.
// You might want to look at Arcus.Security.Provides.AzureKeyVault to ensure this value can be stored in Azure Key Vault instead
string clientSecret = "my-client-secret";
// The name of the resource group containing the Logic App you want to control
string resourceGroup = "my-resource-group";
// The name of the Logic App you want to control.
string logicAppName = "my-logic-app-name";

// Use the LogicAppAuthentication-class to retrieve the required access token.
var authentication = LogicAppAuthentication.UsingServicePrincipal(tenantId, subscriptionId, clientId, clientSecret);
// Once authenticated, use this token to get access to your Logic Apps.

> using (var logicApp = await LogicAppClient.CreateAsync(resourceGroup, logicAppName, authentication))

Reviewer: Maybe remove this line as we explain it later on in a dedicated section?

Author: You are right!

{
}
```

*Want to read more about how to authenticate and gain access to your Logic Apps?*  
*Have a look on [this page](https://invictus-integration.github.io/testing-framework/#/logic-apps/authentication).*  

### Controlling an Azure Logic App

> One of the first and most important aspects of this framework is the ability to have control over every single Logic App that is part of your interface, only by specifying the name of the resource group and the targeted Logic App. Of course, you also need to authenticate, for which you will need to provide the authentication-token which has been retrieved by the previously mentioned operation.  

Reviewer 2: ```suggestion
One of the key aspects of our framework is the ability to have control over every single Azure Logic App that is part of your interface. This can be done by only specifying the name of the resource group and the Azure Logic App itself.
```

> But, once the **LogicAppClient**-object has been created, you can perform any of these operations:  

Reviewer 2: In this post, I'd consider the client technical implementation detail maybe?
```suggestion
Once authenticated, you can perform any of the following operations:
```

> - Create a Logic App

Reviewer 2: Create a new Logic App run or instance?

Author: Not sure why this was added, we don't create logic apps..
Going to remove this.

- Get the Logic App metadata  
- Get the Logic App trigger-URL  
- (Temporarily) enable the Logic App  
- (Temporarily) update the Logic App definition  
- (Temporarily) enable static results on specific actions within the Logic App  
- Trigger a Logic App run  

> - Delete a Logic App  

Reviewer 2: Instance I presume?

Author: And now I'm wondering, why do we offer the possibility to delete a Logic App, but not create one.
Because, this is really about the deletion of an actual logic app.

Author: Feature-request: #57

> All of the above operations should provide sufficient possibilities for you to adjust and trigger your Logic Apps to allow for a specific scenario to be tested.

Reviewer 2: ```suggestion
All of these features provide you the possibility to adjust and trigger your Azure Logic Apps to allow for a specific scenario to be tested.
```

> Below is a small example of how you can enable a Logic App during the execution of your test:

Reviewer 2: ```suggestion
Here is an example of how you can temporarily disable an Azure Logic App during your test:
```

Author: Will be alligned with the below comment.

```csharp
using (var logicApp = await LogicAppClient.CreateAsync(resourceGroup, logicAppName, authentication))
{
    // Any action to be performed before enabling the Logic App.

>     await using (await logicApp.TemporaryEnableAsync())

Reviewer 2: Shouldn't this be disable?

Author: Can go either way - on DEV for instance, you don't want your logic app's to be enabled all the time, so you want to enable it to run your tests and disable it afterwords.

Author: Ok, might be confusing to others, so I agree on that.
Regarding the await statement (it's a copy from the docs 🙉) but basicly it's asynchronously going to dispose of the instance performing the action. (IAsyncDisposable vs IDisposable)

Author: Aha, scratch the previous bit, we don't have a TemporaryDisableAsync()-method, as we didn't see the need for this.
So, I'll add some context on the TemporaryEnableAsync-operation.

> As you don't want your unit/integration test related Logic App to inflict costs while not being used, here is an example of how you can temporarily enable an Azure Logic App during your test:

Reviewer 2: Thanks for checking!

Tracking in #58

    {
        // Perform actions related to your test-case, while the Logic App is enabled.
    }
    // Any action to be performed after disabling the Logic App.
}
```

*Want to read more about how to control a Logic App?*  

> *Have a look on [this page](https://invictus-integration.github.io/testing-framework/#/logic-apps/control-single-logicapp).*  

Reviewer 2: ```suggestion
*Read more about it in [our documentation](https://invictus-integration.github.io/testing-framework/#/logic-apps/control-single-logicapp).*  
```

Think it's ok to put on the line above as well.

```csharp
LogicAppRun logicAppRun =
    await LogicAppsProvider.LocatedAt(resourceGroup, logicAppName, authentication)
                           .WithTimeout(TimeSpan.FromSeconds(15))
                           .WithCorrelationId("08586073923413753771945113291CU110")
                           .PollForSingleLogicAppRunAsync();
```

Next, let's try and retrieve all Logic App runs - we're expecting to see 3 of them - within a timeframe of 60 seconds, all of whom should be containing the tracked property _OrderNumber_ matching the value _123456_:
```csharp
IEnumerable<LogicAppRun> logicAppRuns =
    await LogicAppsProvider.LocatedAt(resourceGroup, logicAppName, authentication)
                           .WithTimeout(TimeSpan.FromSeconds(60))
                           .WithTrackedProperty("OrderNumber", "123456")
                           .PollForLogicAppRunsAsync(minimumNumberOfItems: 3);
```

*Want to read more about how to monitor a Logic App?*  
*Have a look on [this page](https://invictus-integration.github.io/testing-framework/#/logic-apps/polling-logicapp-runs).*  

> Hoping that this will enable and simplify the process of including test-cases within every single integration project on Azure Logic Apps, we are looking forward to receiving feedback allowing us to keep improving and extending our Invictus Test Framework for Azure Logic Apps.

Reviewer: Maybe place this under a 'conclusion' section?

```suggestion
# Conclusion
Hoping that this will enable and simplify the process of including test-cases within every single integration project on Azure Logic Apps, we are looking forward to receiving feedback allowing us to keep improving and extending our Invictus Test Framework for Azure Logic Apps.
```

Reviewer 2: Agreed

> Hoping that this will enable and simplify the process of including test-cases within every single integration project on Azure Logic Apps, we are looking forward to receiving feedback allowing us to keep improving and extending our Invictus Test Framework for Azure Logic Apps.

Reviewer 2: > We are happy to help the Azure Logic Apps community to build automated test suits on top of Azure Logic Apps! We welcome everybody to add feature requests [on GitHub](https://github.com/invictus-integration/testing-framework/) and accept contributions to make the framework even better!

What do you think of this? Yours is fine as well

Author: Yours is better so: yes, master ;)

Since developing the Invictus Methodology, which includes patterns, best practices and several templates, Codit has been investing in building a solid foundation enabling us to quickly kick-off new projects, while ensuring a high level of quality. The biggest thing which, until now, was still lacking, was the ability to include test-cases into those projects.  
Because of this, we started working on a framework that allowed us to easily enable/disable, trigger and even modify existing Azure Logic Apps. But, since the operations are not enough to write full-blown test-scenarios, we extended this list to ensure you would also be able to monitor the executed Azure Logic App runs.  

All these operations have now been forged into the Invictus Test Framework for Azure Logic Apps which we used internally for our customers and have proven to be effective.  
Today, **we are happy to announce that we are open-sourcing Invictus Test Framework for Azure Logic Apps on GitHub** and it is now available on NuGet for you to use on your projects!  

Get started very easily :  
```shell
> Install-Package Invictus.Testing.LogicApps
```

### Authentication

> When you want to use this package to track of control your Logic Apps, you will have to authenticate against the subscription containing said Logic Apps of course.  

Reviewer 2: ```suggestion
When you want to use this package to track of control your Logic Apps, you will have to authenticate against your Azure subscription containing said Logic Apps.  
```

> This can be done by re-using the existing service principal, which has been created to connect your Azure DevOps project to the Azure subscription. However, seeing as this service principal typically has contributor-access to the entire subscription, which is quite an overkill for testing-purposes, it would be advised to create a new service principal, entirely dedicated to the testing scenarios.  

Reviewer 2: How about this? Allows us to tell what they need first and how they can re-use what they have with a warning.

```suggestion

As of today, we provide support for using service principal authentication! Our framework requires to have at least [*Logic App Contributor*](https://docs.microsoft.com/en-us/azure/role-based-access-control/built-in-roles#logic-app-contributor) which gives enough permissions to do our magic.

Of course, you can re-use existing service principals that you already have in Azure DevOps for example. However, seeing as this service principal typically has contributor-access to the entire subscription, which is quite an overkill for testing-purposes, it would be advised to create a new service principal, entirely dedicated to the testing scenarios.  
```

> This test-service principal will need to get the "*Logic App Contributor*"-role assigned, as any other role would be insufficient to update your Logic Apps (to enable static results, for instance).  

Reviewer 2: See above proposal

But what exactly do you need to specify in order to really authenticate?  
Well, below is an overview of the required information:

```csharp

> // The Tenant ID as found in the AAD properties

Reviewer 2: ```suggestion
// The tenant ID of the Azure AD
```

string tenantId = "my-tenant-id";

> // The ID of the subscription containing your Logic Apps.

Reviewer 2: ```suggestion
// The ID of the subscription containing your Azure Logic App.
```

string subscriptionId = "my-subscription-id";

> // The Client ID of the service principal to be used to authenticate.

Reviewer 2: ```suggestion
// The application id of the service principal to be used to authenticate.
```

Reviewer 2: Or am I the only one never knowing if it's the app or object id? :p

Author: Argh, constantly mixing them up.

string clientId = "my-client-id";

> // The Client Secret created along with the service principal used for authentication.

Reviewer 2: ```suggestion
// The client secret created along with the service principal used for authentication.
```

> // You might want to look at Arcus.Security.Provides.AzureKeyVault to ensure this value can be stored in Azure Key Vault instead

Reviewer 2: ```suggestion
// You might want to look at Arcus Security to ensure this value can be stored in Azure Key Vault instead
```

Author: Right, you told me about this earlier...

string clientSecret = "my-client-secret";
// The name of the resource group containing the Logic App you want to control

> string resourceGroup = "my-resource-group";

Reviewer 2: No longer needed in this sample

// The name of the Logic App you want to control.

> string logicAppName = "my-logic-app-name";

Reviewer 2: No longer needed in this sample

Here is an example of how you can temporarily enable an Azure Logic App during your test:  
```csharp
using (var logicApp = await LogicAppClient.CreateAsync(resourceGroup, logicAppName, authentication))
{
    // Any action to be performed before enabling the Logic App.
    await using (await logicApp.TemporaryEnableAsync())
    {
        // Perform actions related to your test-case, while the Logic App is enabled.
    }
    // Any action to be performed after disabling the Logic App.
}
```

*Want to read more about how to control a Logic App?*  
*Read more about it in [our documentation](https://invictus-integration.github.io/testing-framework/#/logic-apps/control-single-logicapp).*  

### Monitoring an Azure Logic App

After preparing the interface for a specific scenario and actually triggering the test, it is of course as important to be able to retrieve enough information to define the outcome of the test.  
Which is why, in addition to the operations listed above you are also able to:  
- Get a Logic App run by 'correlation ID' a.k.a. the 'client tracking ID'
- Get a Logic App run by tracked properties (1 or many)

> - Get a Logic App run within a specific timeframe

Author: Can be both actually.

> - Get a Logic App run within a specific timeframe

Author: ```suggestion
- Get an Azure Logic App run within a specific timeframe  
- Get a list of Azure Logic App runs within a specific timeframe
```

> - Get a Logic App run within a specific timeframe

Reviewer 2: Let's add both then!

> - Get a Logic App run within a specific timeframe

Author: done

> - Get a given number of Logic App runs once completed (polling).   

Author: Well, by completed I mean both in case the LA failed or ran successfully.
But then maybe completed is a bad choice of words?

> - Get a given number of Logic App runs once completed (polling).   

Reviewer 2: Does this better describe it?
```suggestion
- Get a given number of completed Azure Logic App runs (polling) which can be success or failed.   
```

> - Get a given number of Logic App runs once completed (polling).   

Author: Ok, adding this to ensure it's clear to everyone.

  *In case you have the same Logic App running multiple times in parallel.* 

> When we talk about getting a Logic App run, as mentioned above, this means you will be able to access all metadata of this run, including every possible piece of information, such as the status of the executed actions along with the tracked properties per action as well as the global overview of these properties.

Reviewer 2: Trying to condense a bit but too much good stuff in the info!
```suggestion
Every Azure Logic App run provides information about all metadata of the run itself such the status, what triggered it, start/end time, correlation, information & tracked properties for every single action and more!
```

To give you an idea on how you can use this framework to monitor Logic App runs, some small code-samples have been included.  
In this first example we will poll for 15 seconds and look for a single run where the _Correlation Id_ matches a specific value:
```csharp

> LogicAppRun logicAppRun =

Author: I believe it will throw a `TimeOutException` if nothing could be found in the given timeframe.

> LogicAppRun logicAppRun =

Reviewer 2: Ok, we might want to note that then

> LogicAppRun logicAppRun =

Author: done

    await LogicAppsProvider.LocatedAt(resourceGroup, logicAppName, authentication)
                           .WithTimeout(TimeSpan.FromSeconds(15))
                           .WithCorrelationId("08586073923413753771945113291CU110")
                           .PollForSingleLogicAppRunAsync();

> ```

Author: Added the following instead:
> If the framework was unable to find a run within the specific time frame, a `TimeOutException` will be thrown.

> ```

Reviewer 2: Thanks!

Next, let's try and retrieve all Logic App runs - we're expecting to see 3 of them - within a timeframe of 60 seconds, all of whom should be containing the tracked property _OrderNumber_ matching the value _123456_:
```csharp
IEnumerable<LogicAppRun> logicAppRuns =
    await LogicAppsProvider.LocatedAt(resourceGroup, logicAppName, authentication)
                           .WithTimeout(TimeSpan.FromSeconds(60))
                           .WithTrackedProperty("OrderNumber", "123456")
                           .PollForLogicAppRunsAsync(minimumNumberOfItems: 3);

> ```

Author: Same here: a `TimeOutException` is thrown.
Unless I'm wrong, of course, then please do tell me.

> ```

Author: Looked it up and confirmed -> added to the post.
